### ISSUE TITLE
Installation issue: gperftools

### ISSUE THREAD

**Issue by Char-Aznable** [LLNL/spack#25832]
### Steps to reproduce the issue

```console
spack install gperftools
```

### Information on your system

* **Spack:** 0.16.2-3914-fbed679dd0
* **Python:** 3.9.6
* **Platform:** linux-ubuntu20.04-skylake
* **Concretizer:** original

### Additional information

Basically, this hits https://github.com/gperftools/gperftools/issues/424 because gperftools isn't configured to link with libunwind. Forcing a LDFLAGS="-L/path/to/libunwind -lunwind' bypasses the issue but it's an ugly solution. gperftools has cmake build now and we should switch the package to cmake build
@albestro  @eschnett  @msimberg  @teonnik

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by msimberg** [comment]
I can try to have a go at this. It looks like CMake support was added in 2.8.1 though, and I don't know if it's possible to support autotools for the older versions while using CMake for the newer versions? I'd be ok with dropping the older versions but I don't know if someone else needs them to stay around? I take it things build fine with CMake but not autotools, right?

Adding the needed LDFLAGS is not terrible either given that there's one and only one dependency.

**Comment by eschnett** [comment]
There's currently a discussion in the Spack community about how to change build systems, i.e. how to support multiple build systems for packages. I don't know whether there's a good answer yet. In the mean time, I'd just limp forward with autotools, and setting the additional flag.

**Comment by msimberg** [comment]
@Char-Aznable just to understand the problem, are you building an older version of `gperftools`? I just tried `spack install gperftools` on spack commit `edb1d75b1b` which installs `gperftools` 2.9.1 and `libunwind` 1.5.0. I'm on Ubuntu 21.04 and it worked.

The issue you've linked was closed as not a bug, so it's not clear to me in what situations `gperftools` isn't being linked to `libunwind`. I know nothing about `gperftools`' build system otherwise.

**Comment by Char-Aznable** [comment]
@msimberg I was also trying to build 2.9.1. To my understanding, libunwind doesn't have dependent environment set up in package.py so gperftools shouldn't be able to pick up the LDFLAGS from spack's libunwind, should it? I examined my build log and certainly didn't see the desired LDFLAGS. `spack` sometimes pick up dependent library from other directories (see https://github.com/spack/spack/issues/18802). Which libunwind was your build of gperftools linked to?

The issue I linked was probably a bug but the OP probably hacked the autotool config to work around it.

**Comment by haampie** [comment]
Notice that you don't see the flags the compiler wrapper injects in the build logs:

```
$ spack build-env gperftools -- env SPACK_TEST_COMMAND=dump-args gcc foo.c | grep unwind
-I/home/user/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.3.0/libunwind-1.5.0-gu3uqvitxk2jr2l34hqboh7o7eye6d63/include
-L/home/user/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.3.0/libunwind-1.5.0-gu3uqvitxk2jr2l34hqboh7o7eye6d63/lib
-Wl,-rpath,/home/user/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.3.0/libunwind-1.5.0-gu3uqvitxk2jr2l34hqboh7o7eye6d63/lib
```

from my build logs `checking for backtrace in -lunwind... yes`, so it works with just `-lunwind`, provided everything goes through the compiler wrapper.

**Comment by msimberg** [comment]
@Char-Aznable just FYI, I'm not doing any work in this direction at the moment. I'd like to understand why things are failing in your case first.

**Comment by Char-Aznable** [comment]
I still see the same build error on a new machine with a fresh spack setup. The `build-env` command @haampie showed above gave:

```
-L/home/aznb/spack/opt/spack/linux-ubuntu22.04-x86_64_v4/clang-homebrew/libunwind-1.6.2-r6fdxqrwhcd4ozgvbb3sdlzottbyvapj/lib
/home/aznb/spack/opt/spack/linux-ubuntu22.04-x86_64_v4/clang-homebrew/libunwind-1.6.2-r6fdxqrwhcd4ozgvbb3sdlzottbyvapj/lib
```
on my system, which lacks the necessary `-lunwind` flag. I am using clang from homebrew as the compiler instead of gcc

**Comment by msimberg** [comment]
@Char-Aznable I'd be happy to add a `-lunwind` flag if that turns out to be required.

Just for comparison, could you check what your install logs say about `unwind`? This is my verbose install log (on nixos, which may make things a bit different... but still): https://gist.github.com/msimberg/1f2c9d40d04b276cb9a2c7a9b117516a. Note that for me `-lunwind` does explicitly end up in the linker command lines (without any patches, of course).

**Comment by Char-Aznable** [comment]
@msimberg I do see `-lunwind` here and there in the log but not at the place right where it needs to be, exactly the reason why we should move to a more maintainable build system. See the attached log here:
[spack-build-out.txt.zip](https://github.com/spack/spack/files/8953002/spack-build-out.txt.zip)

Manually adding `-lunwind` would work around this issue but it would be difficult to keep up with upstream changes. Anyway, thanks for looking at this.

**Comment by msimberg** [comment]
@Char-Aznable yeah, definitely. I just realized that this now exists in the documentation: https://spack.readthedocs.io/en/latest/build_systems/multiplepackage.html#autotools-cmake. I don't know if this is the multi-build system support that @eschnett was referring to but that looks workable. I'll try to get something going before the weekend.

**Comment by msimberg** [comment]
@Char-Aznable could you try out https://github.com/spack/spack/pull/31231? Note that the 2.10 release notes also mention a CMake/libunwind issue being fixed, so maybe try both 2.10 and at least one of 2.8.1 or 2.9.1 (which are now both also built with CMake). If 2.10 works but the others don't we might want to add a conflict for `+libunwind` and the older versions.

**Comment by haampie** [comment]
Didn't realize you fixed this years ago @msimberg :) 

#43506 fixes it now.
